File: Templates/Java/src/main/java/db/InMemoryDatabase.java
package db;

import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import models.Student;

public class InMemoryDatabase implements IDatabase {
    private final Map<Integer, Student> students = new HashMap<>();

    @Override
    public Student createStudent(Student student) {
        // Find the maximum current id and increment it for the new student
        int nextId = !students.isEmpty()
                ? students.keySet().stream().max(Integer::compare).get() + 1
                : 0;
        student.setId(nextId); // Assign the new id
        students.put(student.getId(), student);
        return student;
    }

    @Override
    public Student getStudent(int id) {
        return students.get(id);
    }

    @Override
    public void updateStudent(Student student) {
        students.put(student.getId(), student);
    }

    @Override
    public void deleteStudent(int id) {
        students.remove(id);
    }

    @Override
    public List<Student> getAllStudents() {
        return students.values().stream().collect(Collectors.toList());
    }
}

File: Templates/Kotlin/src/main/kotlin/db/InMemoryDatabase.kt
package db

import models.Student

class InMemoryDatabase : IDatabase {
    private val students = mutableMapOf<Int, Student>()

    override fun createStudent(student: Student): Student {
        val nextId = if (students.isNotEmpty()) students.keys.maxOrNull()!! + 1 else 0
        student.id = nextId
        students[student.id] = student
        return student
    }

    override fun getStudent(id: Int): Student? {
        return students[id]
    }

    override fun updateStudent(student: Student) {
        students[student.id] = student
    }

    override fun deleteStudent(id: Int) {
        students.remove(id)
    }

    override fun getAllStudents(): List<Student> {
        return students.values.toList()
    }
}
